Zcond final review squashed

MarkoVlaic opened the pull request:
Evo squashano

MarkoVlaic committed:
apply patch

MarkoVlaic committed:
refactor to use zcond

MarkoVlaic committed:
add missing include

MarkoVlaic committed:
cast to a type, not variable

MarkoVlaic committed:
add sdt_probe6_func_t

MarkoVlaic committed:
remove semicolon from macro end

MarkoVlaic committed:
remove sdt_update

MarkoVlaic committed:
SDT_PROBE() proxy for _SDT_PROBE()

MarkoVlaic committed:
typo

MarkoVlaic committed:
netinet build error

MarkoVlaic committed:
remove vm includes from sys/zcond.h

MarkoVlaic committed:
add sys/amd64/amd64/zcond_machdep.c to files.amd64

MarkoVlaic committed:
include queue.h

MarkoVlaic committed:
invtlb()

MarkoVlaic committed:
pmap_invalidate_page()

MarkoVlaic committed:
argument change

MarkoVlaic committed:
pmap_invalidate_page on page boundary

MarkoVlaic committed:
typo

robn committed:
Linux 6.11: add compat macro for page_mapping()

Since the change to folios it has just been a wrapper anyway. Linux has
removed their wrapper, so we add one.

Reviewed-by: Tony Hutter <hutter2@llnl.gov>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Rob Norris <robn@despairlabs.com>
Sponsored-by: https://despairlabs.com/sponsor/
Closes #16400

MarkoVlaic committed:
apply pmap_qenter_pmap() patch

MarkoVlaic committed:
actually call the probe func

MarkoVlaic committed:
supply probe function as _SDT_PROBE() macro argument

MarkoVlaic committed:
adjust argument number

MarkoVlaic committed:
correct wiring of sdt_probe_func

MarkoVlaic committed:
adjust stub argument count

MarkoVlaic committed:
correct sdt_probe6_func cast

MarkoVlaic committed:
get address of stub

MarkoVlaic committed:
change sdt_probe_stub prototype

MarkoVlaic committed:
cast SDT_PROBE6 arguments to uintptr_t

MarkoVlaic committed:
uncomment pstate in netinet/tcp_subr.c

MarkoVlaic committed:
remove dead sdt code

robn committed:
Linux 6.11: avoid passing "end" sentinel to register_sysctl()

Reviewed-by: Tony Hutter <hutter2@llnl.gov>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Rob Norris <robn@despairlabs.com>
Sponsored-by: https://despairlabs.com/sponsor/
Closes #16400

MarkoVlaic committed:
move invltlb() into zcond_after_rendezvous()

MarkoVlaic committed:
squash sdt patch

MarkoVlaic committed:
Merge branch 'sdt_retry' of github.com:MarkoVlaic/freebsd-src into sdt_retry

MarkoVlaic committed:
single out ins_point loading into separate function

MarkoVlaic committed:
rename variable

MarkoVlaic committed:
build error

MarkoVlaic committed:
build error

MarkoVlaic committed:
register event handler for kld

MarkoVlaic committed:
add a printf notification

MarkoVlaic committed:
build error

MarkoVlaic committed:
correct cb signature

MarkoVlaic committed:
printf debugging

MarkoVlaic committed:
squash kld changes

MarkoVlaic committed:
Merge branch 'sdt_retry' of github.com:MarkoVlaic/freebsd-src into sdt_retry

MarkoVlaic committed:
load ins points with linker_file_lookup_symbol

MarkoVlaic committed:
add missing deps argument

MarkoVlaic committed:
add missing entry variable

MarkoVlaic committed:
init entry_sizer

MarkoVlaic committed:
use liner_file_lookup_set for SYSINIT

MarkoVlaic committed:
dereference ins_p

MarkoVlaic committed:
zcond SYSINIT last after SI_SUB_KLD

MarkoVlaic committed:
zcond SYSINIT first after SI_SUB_KLD

MarkoVlaic committed:
last SUB order ANY

MarkoVlaic committed:
printf debugging

MarkoVlaic committed:
unused variable

MarkoVlaic committed:
cast

MarkoVlaic committed:
skip NULL ins points

MarkoVlaic committed:
uncomment owning_zcond

MarkoVlaic committed:
debugging

MarkoVlaic committed:
pinrtf

MarkoVlaic committed:
use owning zcond

MarkoVlaic committed:
check if ins_p is NULL

MarkoVlaic committed:
test everything

MarkoVlaic committed:
check for null zcond

MarkoVlaic committed:
semicolon!!

MarkoVlaic committed:
move printf after owning_zcond calculation

MarkoVlaic committed:
comment out list manipulation

MarkoVlaic committed:
use single pointers for section traversal

MarkoVlaic committed:
remove ins_p dereference

MarkoVlaic committed:
comment out begin/end printf

MarkoVlaic committed:
change sysinit order

MarkoVlaic committed:
subsystem reoredering

MarkoVlaic committed:
change printf

MarkoVlaic committed:
printf on kldload

MarkoVlaic committed:
change sysinit order once again

MarkoVlaic committed:
sysinit order experiment

MarkoVlaic committed:
SI_ORDER_ANY instead of SI_ORDER_LAST

MarkoVlaic committed:
load zcond last after SI_SUB_KLD

MarkoVlaic committed:
load zcond as dtrace subsystem

MarkoVlaic committed:
sysinit first after SI_SUB_KLD

MarkoVlaic committed:
change eventhandlerregister and linker_foreach order

MarkoVlaic committed:
sysinit order again

MarkoVlaic committed:
printf debug

MarkoVlaic committed:
cast printf argument

MarkoVlaic committed:
more printf

MarkoVlaic committed:
hail mary, edit linker script

MarkoVlaic committed:
add start and stop set

MarkoVlaic committed:
refactor

MarkoVlaic committed:
build error

MarkoVlaic committed:
use the correct start/stop symbol

MarkoVlaic committed:
squash ins_point loading with linker_sets

MarkoVlaic committed:
Merge branch 'sdt_refactored' of github.com:MarkoVlaic/freebsd-src into sdt_refactored

MarkoVlaic committed:
all changes since project start

MarkoVlaic committed:
Merge branch 'sdt_refactored' of github.com:MarkoVlaic/freebsd-src into sdt_refactored

MarkoVlaic committed:
add a refcount to zcond

MarkoVlaic committed:
remove unneccessary argument from ZCOND_INIT

MarkoVlaic committed:
add missing include

MarkoVlaic committed:
define and init arg struct

MarkoVlaic committed:
typo

MarkoVlaic committed:
init refcount correctly

MarkoVlaic committed:
printf refcount

MarkoVlaic committed:
printf refcount wrapped

MarkoVlaic committed:
missing comma

MarkoVlaic committed:
i am confused

MarkoVlaic committed:
handle refcount when disable is callend on an already false zcond

MarkoVlaic committed:
remove unused variable

MarkoVlaic committed:
formatting and uint8_t

MarkoVlaic committed:
rename refcount -> refcnt

MarkoVlaic committed:
remove refcount printf

MarkoVlaic committed:
get rid of static cr3 in amd64/zcond_machdep.c

MarkoVlaic committed:
move zcond_pmap handling to machine/pmap.c

MarkoVlaic committed:
refactor kern_zcond.c to use correct number of arguments in pmap_zcond calls. Remove zcond_patching_pmap

MarkoVlaic committed:
take address of zcond_pmap

MarkoVlaic committed:
add unused sysinit parameter

MarkoVlaic committed:
kernel virtual address range

MarkoVlaic committed:
adjust kernel end

MarkoVlaic committed:
ins_point -> patch_point

MarkoVlaic committed:
clang-format

MarkoVlaic committed:
squash review corrections

MarkoVlaic committed:
Merge branch 'sdt_refactored' of github.com:MarkoVlaic/freebsd-src into sdt_refactored

MarkoVlaic committed:
generate random kva address range at boot time

MarkoVlaic committed:
complete function signature

MarkoVlaic committed:
take address of zcond_pmap

MarkoVlaic committed:
remove lock

MarkoVlaic committed:
sysinit zcond SI_SUB after SI_SUB_RANDOM

MarkoVlaic committed:
squash first attempt at zcond kva allocation

MarkoVlaic committed:
Merge branch 'sdt_refactored' of github.com:MarkoVlaic/freebsd-src into sdt_refactored

MarkoVlaic committed:
pmap_update_pde_invalidate

MarkoVlaic committed:
dereference pde

MarkoVlaic committed:
correctly calculate random va to be aligned to 2MB

MarkoVlaic committed:
use vmem_alloc

MarkoVlaic committed:
take addr of

MarkoVlaic committed:
add M_FIRSTFIT flag

MarkoVlaic committed:
remove kva_free

MarkoVlaic committed:
squash second attempt at zcond kva alloc

MarkoVlaic committed:
Merge branch 'sdt_refactored' of github.com:MarkoVlaic/freebsd-src into sdt_refactored

MarkoVlaic committed:
increase zcond va range

MarkoVlaic committed:
patching with a fixed mirror address

MarkoVlaic committed:
full prototype for pmap_qremove_zcond

MarkoVlaic committed:
move patch_page to patch function

MarkoVlaic committed:
use PAGE size with kva_free

MarkoVlaic committed:
typo

MarkoVlaic committed:
squash patching with a single mirror address

MarkoVlaic committed:
Merge branch 'sdt_refactored' of github.com:MarkoVlaic/freebsd-src into sdt_refactored

MarkoVlaic committed:
try to use pmap_qenter_zcond

MarkoVlaic committed:
pmap_qremove_zcond after rendezvous

MarkoVlaic committed:
use pmap_enter

MarkoVlaic committed:
pmap_invalidate_range on zcond_pmap

kprovost committed:
pf: invert direction for inner icmp state lookups

(e.g. traceroute with icmp)
ok henning, jsing

Also extend the test case to cover this scenario.

PR:		280701
Obtained from:	OpenBSD
MFC after:	1 week
Sponsored by:	Rubicon Communications, LLC ("Netgate")

MarkoVlaic committed:
move zcond_before_rendezvous into rendezvous_setup and zcond_after_rendezvous into rendezvous_teardown

MarkoVlaic committed:
move zcond_before_rendezvous into rendezvous_setup and zcond_after_rendezvous into rendezvous_teardown

MarkoVlaic committed:
missing comma

MarkoVlaic committed:
missing comma

MarkoVlaic committed:
perform smp_rendezvous setup and teardown

MarkoVlaic committed:
perform smp_rendezvous setup and teardown

MarkoVlaic committed:
Merge branch 'sdt_refactored' of github.com:MarkoVlaic/freebsd-src into sdt_refactored

MarkoVlaic committed:
Merge branch 'sdt_refactored' of github.com:MarkoVlaic/freebsd-src into sdt_refactored

MarkoVlaic committed:
add a comment and remove mirror_addr from patch_point

MarkoVlaic committed:
add a comment and remove mirror_addr from patch_point

MarkoVlaic committed:
remove accidental file

MarkoVlaic committed:
remove accidental file

MarkoVlaic committed:
remove printf

MarkoVlaic committed:
remove printf

MarkoVlaic committed:
remove unused variable

MarkoVlaic committed:
remove unused variable

ihoro committed:
Add dummymbuf module for testing purposes

Reviewed by:	kp
Differential Revision:	https://reviews.freebsd.org/D45928

ihoro committed:
pf tests: Add 'mbuf' test for (*m0)->m_len < sizeof(struct ip) cases

Reviewed by:	kp
Differential Revision:	https://reviews.freebsd.org/D45927

MarkoVlaic committed:
enclose ZCOND_SET_START_STOP in do/while

MarkoVlaic committed:
enclose ZCOND_SET_START_STOP in do/while

MarkoVlaic committed:
remove __zcond_table from ldscripts

MarkoVlaic committed:
remove __zcond_table from ldscripts

bsdimp committed:
UPDATING: Add note about DMAR

DMAR crashes my laptop on boot, I had to disable it in the boot
laoder. Add a note until this is handled more gracefully.

Sponsored by:		Netflix

wosch committed:
ports(7): searchable index of all ports is at ports.freebsd.org

PR: 280816

bnovkov commented on sys/amd64/amd64/zcond_machdep.c:
```diff
@@ -0,0 +1,104 @@
+	struct pmap zcond_pmap;
+
+	ctxt->cr3 = rcr3();
+	pmap_zcond_get_pmap(&zcond_pmap);
```
Malo sam gledao ostatak `amd64/amd64/` koda, ništa nas ne sprječava da tablicom manipuliramo direktno iz `zcond_machdep.c` (`<vm/pmap.h>` includea i MD pmap dijelove).

Predlažem da sve PMAP promjene premjestiš tu, tako će biti najčišće, bez potrebe za dodavanjem novih sučelja u `pmap.h`.

bnovkov commented on sys/amd64/amd64/zcond_machdep.c:
```diff
@@ -0,0 +1,104 @@
+insn_nop(uint8_t insn[], size_t size)
+{
+	int i;
+	if (size == ZCOND_INSN_SHORT_SIZE) {
```
```suggestion

	if (size == ZCOND_INSN_SHORT_SIZE) {
```

bnovkov commented on sys/amd64/amd64/zcond_machdep.c:
```diff
@@ -0,0 +1,104 @@
+insn_jmp(uint8_t insn[], size_t size, vm_offset_t offset)
+{
+	int i;
+	if (size == 2) {
```
```suggestion

	if (size == 2) {
```

bnovkov replied:
Magičan broj.

bnovkov commented on sys/amd64/amd64/zcond_machdep.c:
```diff
@@ -0,0 +1,104 @@
+{
+	uint8_t *patch_addr = (uint8_t *)p->patch_addr;
+	vm_offset_t offset;
+
```
```suggestion
	uint8_t *patch_addr;
	vm_offset_t offset;
       
        patch_addr = (uint8_t *)p->patch_addr;
```

bnovkov commented on sys/amd64/amd64/zcond_machdep.c:
```diff
@@ -0,0 +1,104 @@
+	vm_offset_t offset;
+
+	if (*patch_addr == nop_short_bytes[0]) {
+		// two byte nop
```
```suggestion
		/* two byte nop */
```

bnovkov replied:
Isto za sve `//` komentare.

bnovkov commented on sys/amd64/include/zcond.h:
```diff
@@ -0,0 +1,34 @@
+/* from Intel® 64 and IA-32 Architectures Software Developer’s Manual, Volume 2B
+ * 4-165 */
+static char nop_short_bytes[] = { 0x66, 0x90 };
+static char nop_long_bytes[] = { 0x0f, 0x1f, 0x44, 0x00, 0x00 };
```
```suggestion
static uint8_t nop_short_bytes[] = { 0x66, 0x90 };
static uint8_t nop_long_bytes[] = { 0x0f, 0x1f, 0x44, 0x00, 0x00 };
```

bnovkov commented on sys/cddl/dev/sdt/sdt.c:
```diff
@@ -494,7 +428,8 @@ sdt_load(void)
 
-	sdt_probe_func = sdt_dtrace_probe;
+	sdt_probe_func = dtrace_probe;
+    sdt_probe6_func = (sdt_probe6_func_t)sdt_dtrace_probe;
```
Formatting.

bnovkov commented on sys/cddl/dev/sdt/sdt.c:
```diff
@@ -293,7 +245,7 @@ sdt_disable(void *arg __unused, dtrace_id_t id, void *parg)
 	KASSERT(probe->sdtp_lf->nenabled > 0, ("no probes enabled"));
 
-	sdt_probe_update(probe, false);
+    zcond_disable(probe->enabled);
```
Formatting.

bnovkov commented on sys/kern/kern_zcond.c:
```diff
@@ -0,0 +1,358 @@
+}
+SYSINIT(zcond, SI_SUB_ZCOND, SI_ORDER_FIRST, zcond_init, NULL);
+
+struct rendezvous_data {
```
```suggestion
struct zcond_patch_arg {
```

bnovkov commented on sys/kern/kern_zcond.c:
```diff
@@ -0,0 +1,358 @@
+static void
+rendezvous_action(void *arg)
+{
+	struct rendezvous_data *data;
+	data = (struct rendezvous_data *)arg;
+
```
```suggestion
	struct rendezvous_data *data;
	
	data = (struct rendezvous_data *)arg;
```

bnovkov replied:
Ista stvar u funkciji ispod.

bnovkov commented on sys/kern/kern_zcond.c:
```diff
@@ -0,0 +1,358 @@
+{
+	struct rendezvous_data *data;
+	data = (struct rendezvous_data *)arg;
+
```
```suggestion
	struct rendezvous_data *data;
	
	data = (struct rendezvous_data *)arg;
```

bnovkov commented on sys/kern/kern_zcond.c:
```diff
@@ -0,0 +1,358 @@
+}
+
+void
+__zcond_set_enabled(struct zcond *cond, bool new_state)
```
```suggestion
__zcond_toggle(struct zcond *cond, bool enable)
```

bnovkov commented on sys/amd64/amd64/zcond_machdep.c:
```diff
@@ -0,0 +1,104 @@
+insn_nop(uint8_t insn[], size_t size)
+{
+	int i;
+	if (size == ZCOND_INSN_SHORT_SIZE) {
```
Jesu li ovo compile-time konstante stvari? Tj,  hoće li `insn_nop` i `insn_jmp` stvarno primati i `SHORT_SIZE` i `LONG_SIZE` tokom rada?

MarkoVlaic replied:
Je compile-time u smislu da se kod prevodi ili u kratki ili dugi `jmp`. Međutim tu informaciju ne mogu zapisati u elf section ili barem ne znam kako bih. 
Iz `zcond_get_patch_insn` se prosljeduje `SHORT_SIZE` ili `LONG_SIZE` ovisno o tipu instrukcije na koji se naleti.

bnovkov commented on sys/amd64/amd64/zcond_machdep.c:
```diff
@@ -0,0 +1,104 @@
+}
+
+void
+zcond_get_patch_insn(struct patch_point *p, uint8_t insn[], size_t *size)
```
Koliko shvaćam, ovdje gledaš jel na adresi trenutno NOP ili JMP pa onda biraš što ćeš upržiti?
Može li se izbjeći ove konstrukte s labelama ako funkciji predaš argument `bool enable`?

MarkoVlaic replied:
Kada bih slao `bool enable` mogao bih izbjeći provjeru je li na adresi `NOP` ILI `JMP`, međutim i dalje moram odrediti radi li se o dugoj ili kratkoj instrukciji. 
Iz tog razloga mislim da ne bismo bas puno toga riješili uvođenjem `bool enable`.

bnovkov commented on sys/sys/zcond.h:
```diff
@@ -0,0 +1,243 @@
+ * Bakes in a jmp instruction instruction, so the return value is initially
+ * true.
+ */
+static __attribute__((always_inline)) bool
```
```suggestion
static __always_inline bool
```

bnovkov commented on sys/sys/zcond.h:
```diff
@@ -0,0 +1,243 @@
+ * Bakes in a nop instruction instruction, so the return value is initially
+ * false.
+ */
+static __attribute__((always_inline)) bool
```
```suggestion
static __always_inline bool
```

bnovkov commented on sys/sys/zcond.h:
```diff
@@ -0,0 +1,243 @@
+ * A single optimized boolean.
+ */
+struct zcond {
+	bool enabled;
```
```suggestion
```

Refcount je tu jedini mjerodavan imo, ako je pri enableu `refcount_acquire > 0` znači da je zcond upogonjen ili će vrlo ubrzo biti.

bnovkov commented on sys/sys/sdt.h:
```diff
@@ -452,6 +414,7 @@ struct sdt_probe {
 	id_t		id;		/* DTrace probe ID. */
 	int		n_args;		/* Number of arguments. */
 	struct linker_file *sdtp_lf;	/* Module in which we're defined. */
+    DECLARE_ZCOND_FALSE(enabled);
```
Formatting.

MarkoVlaic replied:
meni je nesto s vimom lose postavljeno jer se kod mene u editoru ovo prikazuje normalno.. Tako i za ostale formatting greske

bnovkov commented on sys/sys/sdt.h:
```diff
@@ -177,6 +178,7 @@ SET_DECLARE(sdt_argtypes_set, struct sdt_argtype);
 		    .mod = #_mod,					\
 		    .func = #_func,					\
 		    .name = #_name,					\
+            .enabled = ZCOND_INIT(false)    \
```
Formatting.

bnovkov commented on sys/sys/kernel.h:
```diff
@@ -107,6 +107,7 @@ enum sysinit_sub_id {
 	SI_SUB_RACCT		= 0x2110000,	/* resource accounting */
 	SI_SUB_KDTRACE		= 0x2140000,	/* Kernel dtrace hooks */
 	SI_SUB_RANDOM		= 0x2160000,	/* random number generator */
+    SI_SUB_ZCOND    = 0x2170000,
```
Formatting.

bnovkov commented on sys/kern/kern_zcond.c:
```diff
@@ -0,0 +1,358 @@
+	} else {
+		refcount_acquire(&cond->refcnt);
+	}
+
+	if (cond->enabled == new_state) {
+		return;
+	}
```
```suggestion
```

bnovkov commented on sys/kern/kern_zcond.c:
```diff
@@ -0,0 +1,358 @@
+{
+	struct zcond_md_ctxt ctxt;
+
+	if (new_state == false) {
+		if (refcount_release_if_not_last(&cond->refcnt)) {
+			/* refcount > 1 */
+			return;
+		} else if (!refcount_release_if_last(&cond->refcnt)) {
+			/* refcount == 0 */
+			return;
+		}
+	} else {
+		refcount_acquire(&cond->refcnt);
+	}
```
```suggestion
     if (enable && refcount_acquire(&cond->refcnt) > 0)
         return;
     else if (!enable && !refcount_release(&cond->refcnt))
         return;
```

MarkoVlaic replied:
puno elegantnije :)

bnovkov commented on sys/kern/kern_zcond.c:
```diff
@@ -0,0 +1,358 @@
+	smp_rendezvous(rendezvous_setup, rendezvous_action, rendezvous_teardown,
+	    &arg);
+	zcond_after_rendezvous(&ctxt);
+	pmap_qremove_zcond();
```
Staviti u `zcond_after_rendezvous`

bnovkov commented on sys/kern/kern_zcond.c:
```diff
@@ -0,0 +1,358 @@
+	struct patch_point *p;
+	vm_offset_t mirror_addr;
+	vm_page_t patch_page;
+	unsigned char insn[ZCOND_MAX_INSN_SIZE];
```
```suggestion
	uint8_t insn[ZCOND_MAX_INSN_SIZE];
```

bnovkov commented on sys/kern/kern_zcond.c:
```diff
@@ -0,0 +1,358 @@
+zcond_patch(struct zcond *cond, bool new_state)
+{
+	struct patch_point *p;
+	vm_offset_t mirror_addr;
```
Nema potrebe da adresa bude tu, obzirom da je adresa uvijek ista trebalo bi ju samo zapamtiti pri SYSINIT u neku statičku varijablu unutar `zcond_machdep` pa ju dohvatiti s nekom funkcijom.

glebius committed:
zfs: fix build without MAC

bnovkov commented on sys/amd64/amd64/pmap.c:
```diff
@@ -12311,4 +12312,70 @@ DB_SHOW_COMMAND(ptpages, pmap_ptpages)
+
+void
+pmap_zcond_get_pmap(struct pmap* pm) {
+   memcpy(pm, &zcond_pmap, sizeof(zcond_pmap)); 
```
Nema potrebe za kopiranjem strukture ovdje :')

bnovkov commented on sys/amd64/amd64/pmap.c:
```diff
@@ -12311,4 +12312,70 @@ DB_SHOW_COMMAND(ptpages, pmap_ptpages)
+                X86_PG_M | X86_PG_RW | X86_PG_V);
+    }
+
+    if (__predict_false((oldpte & X86_PG_V) != 0))
+            pmap_invalidate_range(&zcond_pmap, zcond_patch_va, zcond_patch_va + PAGE_SIZE);
```
```suggestion
    invplg(zcond_patch_va);
```
Invalidate range će potencijalno okinut novi IPI, što će panicat.

bnovkov commented on sys/amd64/amd64/pmap.c:
```diff
@@ -12311,4 +12312,70 @@ DB_SHOW_COMMAND(ptpages, pmap_ptpages)
+	    kern_end - kern_start, kern_start);
+
+    zcond_patch_va = kva_alloc(PAGE_SIZE);
+    dummy_page = vm_page_alloc_noobj(VM_ALLOC_WIRED);
```
```suggestion
    dummy_page = vm_page_alloc_noobj(VM_ALLOC_WIRED | VM_ALLOC_NOFREE);
```

Zasad, pogledat ću možemo li osloboditi stranicu pa updateam komentar.

bnovkov commented on sys/amd64/amd64/pmap.c:
```diff
@@ -12311,4 +12312,70 @@ DB_SHOW_COMMAND(ptpages, pmap_ptpages)
+    kern_start = virtual_avail;
+    kern_end = kernel_vm_end;
+
+	memset(&zcond_pmap, 0, sizeof(zcond_pmap));
```
Uvući

bnovkov commented on sys/amd64/amd64/pmap.c:
```diff
@@ -12311,4 +12312,70 @@ DB_SHOW_COMMAND(ptpages, pmap_ptpages)
+    pt_entry_t *pte;
+    int cache_bits;
+
+    oldpte = 0;
+    pte = pmap_pte(&zcond_pmap, zcond_patch_va);
```
Predlažem da ovaj lookup odradiš u initu - imamo garanciju da nitko neće dirati taj PTE tako da nema potrebe da ga stalno lookupamo.
Slobodno ga zapamti u sysinitu i zamijeni ovaj lookup.

bnovkov commented on sys/amd64/amd64/pmap.c:
```diff
@@ -12311,4 +12312,70 @@ DB_SHOW_COMMAND(ptpages, pmap_ptpages)
+pmap_qremove_zcond(void) {
+    pt_entry_t *pte;
+
+    pte = pmap_pte(&zcond_pmap, zcond_patch_va);
```
Ista stvar za PTE ko i gore.

bnovkov reviewed (commented):
Ostavio sam komentare, većinom je sve ok ali bit će još samo malo micanja koda okolo, mislim da došli do dobre strukture.

Od većih stvari je samo PMAP komentar - to bi sve trebalo premjestiti u `zcond_machdep.c` i maknuti nova pmap sučelja.

A druga stvar koju sam tek sad primjetio, u novim datotekama svugdje fali copyright clause:
```
/*-
 * SPDX-License-Identifier: BSD-2-Clause
 *
 * Copyright (c) 2024 Marko Vlaić <email>
 *
* This code was developed as a Google Summer of Code 2024. project
 * under the guidance of Bojan Novković <bnovkov@freebsdorg>.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */
```

tonyhutter committed:
ZTS: Use /dev/urandom instead of /dev/random

Use /dev/urandom so we never have to wait on entropy.

Reviewed-by: George Melikov <mail@gmelikov.ru>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Tony Hutter <hutter2@llnl.gov>
Closes #16442

ccfbsd committed:
tcp cc: clean up some un-used cc_var flags

Reviewed by: tuexen
Differential Revision: https://reviews.freebsd.org/D46299

Commit:
Add missing kstats to dataset kstats

Reviewed-by: Tony Nguyen <tony.nguyen@delphix.com>
Reviewed-by: Rob Norris <robn@despairlabs.com>
Signed-off-by: Paul Dagnelie <pcd@delphix.com>
Closes #16431

Commit:
Add missing kstats to dataset kstats

Reviewed-by: Tony Nguyen <tony.nguyen@delphix.com>
Reviewed-by: Rob Norris <robn@despairlabs.com>
Signed-off-by: Paul Dagnelie <pcd@delphix.com>
Closes #16431
(cherry picked from commit 244ea5c4881f92a9d7c1fb341a49b127fda7539d)

jsai20 committed:
Fix projid accounting for xattr objects

zpool upgraded with 'feature@project_quota' needs re-layout of SA's
to fix the SA_ZPL_PROJID at SA_PROJID_OFFSET (128). Its necessary for
the correct accounting of object usage against its projid.
Old object (created before upgrade) when gets a projid assigned, its
SA gets re-layout via sa_add_projid(). If object has xattr dir, SA
of xattr dir also gets re-layout. But SA re-layout of xattr objects
inside a xattr dir is not done.

Fix zfs_setattr_dir() to re-layout SA's on xattr objects, when setting
projid on old xattr object (created before upgrade).

Reviewed-by: Tony Hutter <hutter2@llnl.gov>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Jitendra Patidar <jitendra.patidar@nutanix.com>
Closes #16355
Closes #16356

TuxPowered42 committed:
pf: Fix indentation in struct pf_ksrc_node

This is a purely cosmetic change to simplify future diffs.

Reviewed by:	kp
Differential Revision:	https://reviews.freebsd.org/D46298

TuxPowered42 committed:
pf: Convert struct pf_addr_wrap before sending it over netlink

The struct pf_addr_wrap when used inside of kernel operates on pointers to
tables or interfaces. When reading a ruleset the struct must contain
counters calculated from the aforementioned tables and interfaces. Both the
pointers and the resulting counters are stored in an union and thus can't be
present in the struct at the same time.

The original ioctl code handles this by making a copy of struct pf_addr_wrap
for pool addresses, accessing the table or interface structures by their
pointers, calculating the counter values and storing them in place of those
pointers in the copy. Then this copy is sent over ioctl.

Use this mechanism for netlink too. Create a copy of src/dst addresses. Use
the existing function pf_addr_copyout() to convert pointers to counters both
for src/dst and pool addresses.

Reviewed by:	kp
Differential Revision:	https://reviews.freebsd.org/D46291

kprovost committed:
pf tests: basic test for sticky-address configuration on rules

There was a bug in pf_handle_get_addr() where it confused the counter and
pointer in the pf_addr_wrap.p union, causing panics. Test for this.

Sponsored by:	Rubicon Communications, LLC ("Netgate")

MarkoVlaic committed:
move pmap functionality to zcond_machdep

MarkoVlaic committed:
move pmap functionality to zcond_machdep

MarkoVlaic committed:
missing semicolon

MarkoVlaic committed:
missing semicolon

MarkoVlaic committed:
add some includes

MarkoVlaic committed:
add some includes

MarkoVlaic committed:
more includes

MarkoVlaic committed:
more includes

MarkoVlaic committed:
more includes

MarkoVlaic committed:
more includes

MarkoVlaic committed:
replace pmap_pte with zcond_pte

MarkoVlaic committed:
replace pmap_pte with zcond_pte

MarkoVlaic committed:
get pml4 correctly

MarkoVlaic committed:
get pml4 correctly

MarkoVlaic committed:
is_la57. Remove unused m

MarkoVlaic committed:
is_la57. Remove unused m

mmatuska committed:
zfs: merge openzfs/zfs@d2ccc2155

Notable upstream pull request merges:
 #16431 244ea5c48 Add missing kstats to dataset kstats

Obtained from:	OpenZFS
OpenZFS commit: d2ccc2155217bfdd7e19db4480d981a8f3fad46f

MarkoVlaic committed:
use zcond_pmap for la57 calculation

MarkoVlaic committed:
use zcond_pmap for la57 calculation

MarkoVlaic committed:
squash pmap_pte

MarkoVlaic committed:
squash pmap_pte

MarkoVlaic committed:
Merge branch 'sdt_refactored' of github.com:MarkoVlaic/freebsd-src into sdt_refactored

MarkoVlaic committed:
Merge branch 'sdt_refactored' of github.com:MarkoVlaic/freebsd-src into sdt_refactored

MarkoVlaic committed:
adjust kassert

MarkoVlaic committed:
adjust kassert

MarkoVlaic committed:
!= instead of == in kassert

MarkoVlaic committed:
!= instead of == in kassert

MarkoVlaic committed:
allow pte to be 0

MarkoVlaic committed:
allow pte to be 0

MarkoVlaic committed:
call pmap_pte on sysinit and nowhere else

MarkoVlaic committed:
call pmap_pte on sysinit and nowhere else

MarkoVlaic committed:
invlpg

MarkoVlaic committed:
invlpg

MarkoVlaic committed:
fetch mirror addr at sysinit

MarkoVlaic committed:
fetch mirror addr at sysinit

MarkoVlaic committed:
cosmetic changes mostly

MarkoVlaic committed:
cosmetic changes mostly

MarkoVlaic committed:
semicolon -.-

MarkoVlaic committed:
semicolon -.-

MarkoVlaic committed:
refactor

MarkoVlaic committed:
refactor

MarkoVlaic committed:
remove VM_ALLOC_NOFREE

MarkoVlaic committed:
remove VM_ALLOC_NOFREE

MarkoVlaic committed:
remove zcond.enabled, rely on refcnt instead

MarkoVlaic committed:
remove zcond.enabled, rely on refcnt instead

MarkoVlaic committed:
leave zcond_patch_arg.enable

MarkoVlaic committed:
leave zcond_patch_arg.enable

MarkoVlaic committed:
add enable to struct defintion

MarkoVlaic committed:
add enable to struct defintion

MarkoVlaic committed:
remove kva_free()

MarkoVlaic committed:
remove kva_free()

MarkoVlaic committed:
missing comma

MarkoVlaic committed:
missing comma

markjdb committed:
arm64: Clamp segment sizes properly in bounce_bus_dmamap_load_buffer()

Commit 099b59515499 ("Improve loading of multipage aligned buffers.")
modified bounce_bus_dmamap_load_buffer() with the assumption that busdma
memory allocations are physically contiguous, which is not always true:
bounce_bus_dmamem_alloc() will allocate memory with
kmem_alloc_attr_domainset() in some cases, and this function is not
guaranteed to return contiguous memory.

The damage seems to have been mitigated for most consumers by clamping
the segment size to maxsegsz, but this was removed in commit
a77e1f0f81df ("busdma: better handling of small segment bouncing"); in
practice, it seems busdma memory is often allocated with maxsegsz ==
PAGE_SIZE.  In particular, after commit a77e1f0f81df I see occasional
random kernel memory corruption when benchmarking TCP through mlx5
interfaces.

Fix the problem by using separate flags for contiguous and
non-contiguous busdma memory allocations, and using that to decide
whether to clamp.

Fixes:	099b59515499 ("Improve loading of multipage aligned buffers.")
Fixes:	a77e1f0f81df ("busdma: better handling of small segment bouncing")
Sponsored by:	Klara, Inc.
Sponsored by:	Stormshield
Differential Revision:	https://reviews.freebsd.org/D46238

MarkoVlaic committed:
misleading indentaion -.-

MarkoVlaic committed:
misleading indentaion -.-

MarkoVlaic committed:
avoid refcnt underflow on refcount_release

MarkoVlaic committed:
avoid refcnt underflow on refcount_release

MarkoVlaic committed:
refcount starts at 2

MarkoVlaic committed:
refcount starts at 2

MarkoVlaic committed:
dont panic on disabled zcond disable

MarkoVlaic committed:
dont panic on disabled zcond disable

MarkoVlaic committed:
VM_ALLOC_NOFREE

MarkoVlaic committed:
merge conflicts

MarkoVlaic committed:
remove panic when disabling disabled zcond

MarkoVlaic committed:
clang-format
